### PR TITLE
[jit] make QualifiedName a value type

### PR DESCRIPTION
**Pull request by suo**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #19575 [jit] Use QualifiedName for classes
* **#19626 [jit] make QualifiedName a value type**
* #19566 [jit] separate QualifiedName into its own file

Previously QualifiedName was a recursive reference type, which is more
efficient as it avoids copies. But I want to make it a value type
because:
- People think of names as values anyway
- When doing stuff with QualifiedName I found myself doing a lot of
recursive prefix traversals and copies anyway.

Differential Revision: [D15049582](https://our.internmc.facebook.com/intern/diff/D15049582)

### CODE REVIEW

**Review by zdevito — COMMENTED** [review]
In person: let's anticipate a future when someone will use this class where the names might not be valid identifiers and may contain the standard delimiter. We do not have this case now, but we also do not want to have to make a third qualified name class.

File: aten/src/ATen/core/qualified_name.h [hunk 1]
```diff
@@ -1,91 +1,96 @@
...
+  bool operator!=(const QualifiedName& other) const {
+    return !(*this == other);
   }
 
  private:
-  std::ostream& toString(std::ostream& ss) const {
-    if (prefix_ == nullptr) {
-      ss << name_;
-      return ss;
+  char delimiter_ = '.';
```
Comment (eellison) [inline aten/src/ATen/core/qualified_name.h]: this could be const

File: aten/src/ATen/core/qualified_name.h [hunk 2]
```diff
@@ -1,91 +1,96 @@
...
+// Represents a name of the form "foo.bar.baz"
+struct QualifiedName {
+  QualifiedName() {}
 
-  // Create a qualified name from a dotted string, splitting it as necessary.
-  // Example input: "foo.bar.baz"
-  static QualifiedNamePtr createFromDotted(const std::string& name) {
+  // `name` can be a dotted string, like "foo.bar.baz", or just a bare name.
+  explicit QualifiedName(std::string name) {
+    AT_ASSERT(!name.empty());
```
Comment (eellison) [inline aten/src/ATen/core/qualified_name.h]: nit: this check is unnecessary, and the one below has a better msg

File: aten/src/ATen/core/qualified_name.h [hunk 3]
```diff
@@ -1,91 +1,96 @@
...
-  static QualifiedNamePtr create(QualifiedNamePtr prefix, std::string name) {
-    return c10::make_intrusive<QualifiedName>(
-        std::move(prefix), std::move(name));
-  }
-  static QualifiedNamePtr create(std::string name) {
-    return c10::make_intrusive<QualifiedName>(
-        QualifiedNamePtr(), std::move(name));
-  }
+// Represents a name of the form "foo.bar.baz"
+struct QualifiedName {
```
Comment (eellison) [inline aten/src/ATen/core/qualified_name.h]: Question: why remove QualifiedNamePtr?

**Review by eellison — APPROVED** [review]
Looks fine, but now each Qualified name is taking up twice as much storage, prefix + name_ and qualifiedName_.

File: aten/src/ATen/core/qualified_name.h (hunk 1 above)
Comment (suo) [reply aten/src/ATen/core/qualified_name.h]: making members const deletes their copy constructor, which is weird for something that should be a value type. Having it be private and all methods const accomplishes much the same goal

File: aten/src/ATen/core/qualified_name.h (hunk 3 above)
Comment (suo) [reply aten/src/ATen/core/qualified_name.h]: I explain the reasoning in the PR description

**Pull request closed without merging**